neo-starter-rabbitmq/src/main/java/com/neoteric/starter/rabbit/RetryMessageRecoverer.java

```diff
@@ -0,0 +1,127 @@
+    private volatile String dleExchangePrefix = "dle-";
+    private final AmqpAdmin amqpAdmin;
+    private final String predefinedDleExchange;
+
+    public RetryMessageRecoverer(AmqpTemplate errorTemplate, AmqpAdmin amqpAdmin, String predefinedDleExchange) {
+        this.amqpAdmin = amqpAdmin;
+        Assert.notNull(errorTemplate, "'errorTemplate' cannot be null");
+        this.errorTemplate = errorTemplate;
+        this.predefinedDleExchange = predefinedDleExchange;
+    }
+
+    @Override
+    public void recover(Message message, Throwable cause) {
+        String originalExchange = message.getMessageProperties().getReceivedExchange();
+        String originalQueueName = message.getMessageProperties().getConsumerQueue();
+        String originalRoutingKey = message.getMessageProperties().getReceivedRoutingKey();
+        Map<String, Object> headers = message.getMessageProperties().getHeaders();
+        Long messageCount = getXDeathCount(message);
+        headers.put(X_EXCEPTION_STACKTRACE, getStackTraceAsString(cause));
+        headers.put(X_EXCEPTION_MESSAGE, cause.getCause() != null ? cause.getCause().getMessage() : cause.getMessage());
+        headers.put(X_ORIGINAL_EXCHANGE, originalExchange);
+        headers.put(X_ORIGINAL_ROUTING_KEY, message.getMessageProperties().getReceivedRoutingKey());
+
+        String targetExchange = declareDleExchange(originalExchange);
+        String targetRoutingKey;
+        if (messageCount < 3) {
+            targetRoutingKey = retryRoutingKey(originalRoutingKey);
+            createAdditionalBindedQueue(retryQueue(originalQueueName), targetRoutingKey, targetExchange, ImmutableMap.of(
+                    X_DEAD_LETTER_EXCHANGE, originalExchange,
+                    X_DEAD_LETTER_ROUTING_KEY, originalRoutingKey,
+                    X_MESSAGE_TTL, 900000));
+
+        } else {
+            targetRoutingKey = dleRoutingKey(originalRoutingKey);
+            createAdditionalBindedQueue(dleQueue(originalQueueName), targetRoutingKey, targetExchange, ImmutableMap.of());
+        }
+
+        this.errorTemplate.send(targetExchange, targetRoutingKey, message);
+
+        LOG.info("Republishing failed message to exchange: '{}' with routing key: '{}'. Redelivery nr: {} ",
+                targetExchange, targetRoutingKey, messageCount);
+    }
+
+
+    private Long getXDeathCount(Message message) {
+        List<Object> xDeathHeader = (List<Object>) message.getMessageProperties().getHeaders().get(X_DEATH);
+        if (xDeathHeader == null || xDeathHeader.isEmpty()) {
+            return 0l;
+        }
+        Map<String, Object> xDeathHeaderProperties = (Map<String, Object>) xDeathHeader.get(0);
+        return (Long) xDeathHeaderProperties.getOrDefault(DEATH_COUNT, 0);
+    }
+
+    private String declareDleExchange(String originalExchangeName) {
+        String dleExchangeName = dleExchange(originalExchangeName);
+        DirectExchange exchange = new DirectExchange(dleExchangeName, true, false);
```

Comment:
As the Ampq API is as is, please use 'true' and 'false' constants with meaningful names